pkg/cache/cache.go

```diff
@@ -0,0 +1,283 @@
+
+		if existingObj, found := c.threadSafeCache.Get(key); found {
+
+			glog.Debugf("%#v found in cache. comparing..", existingObj)
+			diff := pretty.Compare(existingObj, newObj)
+			glog.Debugf("Diff: %s", diff)
+
+			if len(diff) != 0 {
+
+				glog.Debugf("%#v and %#v do not match.Updating it in calico cache.", newObj, existingObj)
+
+				c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+				c.workqueue.Add(key)
+			}
+		} else {
+			glog.Debugf("%#v not found in calico cache. Adding it in calico cache", newObj)
+
+			c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+			c.workqueue.Add(key)
+		}
+	} else {
+		glog.Errorf("%#v is not of type %s. Not storing in cache.", newObj, c.ObjectType)
+	}
+}
+
+func (c *calicoCache) Delete(key string) {
+
+	glog.Debug("Deleting %s in calico", key)
+	c.threadSafeCache.Delete(key)
+	c.workqueue.Add(key)
+}
+
+func (c *calicoCache) Get(key string) (interface{}, bool) {
+
+	obj, found := c.threadSafeCache.Get(key)
+	if found {
+		return obj, true
+	}
+	return nil, false
+}
+
+// Prime funtion adds object to threadSafeCache.
+// Only difference with Set() call is it does not queue the key
+// to workqueue. Stores pointer of value object in  cache
+func (c *calicoCache) Prime(key string, value interface{}) {
+
+	c.threadSafeCache.Set(key, &value, cache.NoExpiration)
+}
+
+// ListKeys returns list of calico cache keys in the form
+// of array of strings. Cache stores name field of objects as key.
+func (c *calicoCache) ListKeys() []string {
+
+	cacheItems := c.threadSafeCache.Items()
+	keys := make([]string, 0, len(cacheItems))
+	for k := range cacheItems {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+func (c *calicoCache) SetIndexer(indexer k8sCache.Indexer) {
+	c.k8sCacheIndexer = indexer
+}
+
+func (c *calicoCache) GetQueue() workqueue.RateLimitingInterface {
+	return c.workqueue
+}
+func (c *calicoCache) HasSynced() bool {
+	return c.synced
+}
+
+// Load all the calico datastore objects at the begining of run
+func (c *calicoCache) Run(stopChan chan struct{}) {
+
+	// how should we handled failed priming?
+	// with failed priming calico cache will be empty
+	// but it will still function. Only any manual
+	// additions of objects in ETCD datastore will not
+	// be detected and ETCD datastore get bombarded with
+	// events for all k8s objects.
+	c.primeCache()
+	go c.reconcile(stopChan)
+}
+
+// primeCache() function populates Calico Cache with only calico objects
+// that are created by policy controller.
+func (c *calicoCache) primeCache() error {
+
+	etcdObjList, err := c.ListFunc()
+
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	for _, profile := range etcdObjList {
+		calicoKey := c.KeyFunc(profile)
+		c.Prime(calicoKey, profile)
+	}
+
+	c.synced = true
+	return nil
+}
+
+func (c *calicoCache) reconcile(stopChan chan struct{}) {
+
+	glog.Info("Starting periodic resync thread")
+
+	ticker := time.NewTicker(time.Second * 20)
```

Comment:
nit: We need to make this configurable, and it should default to something like 5m